pkg/machinehandler/machinehandler.go

```diff
@@ -0,0 +1,132 @@
+package machinehandler
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	ErrUnstructuredFieldNotFound = fmt.Errorf("field not found")
+)
+
+type MachineHandler struct {
+	APIGroup string
+	Client   client.Client
+	Config   *rest.Config
+	Ctx      context.Context
+}
+
+type Machine struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Status            MachineStatus `json:"status,omitempty"`
+}
+type MachineStatus struct {
+	NodeRef   *corev1.ObjectReference `json:"nodeRef,omitempty"`
+	Addresses []corev1.NodeAddress    `json:"addresses,omitempty"`
+}
+
+// ListMachines list all machines using given client
+func (m *MachineHandler) ListMachines() ([]Machine, error) {
+	APIVersion, err := m.getAPIGroupPreferredVersion()
+	if err != nil {
+		return nil, err
+	}
+
+	unstructuredMachineList := &unstructured.UnstructuredList{}
+	unstructuredMachineList.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   m.APIGroup,
+		Kind:    "MachineList",
+		Version: APIVersion,
+	})
+	if err := m.Client.List(m.Ctx, unstructuredMachineList); err != nil {
```

Comment:
This is a cluster scoped list for machines. @enxebre  do you think if we'll need namespace scoped list?

Reply:
in my case, CAPI and MAPI machines will exist in different namespaces so we want to capture all machines.

Reply:
> CAPI and MAPI machines will exist in different namespaces so we want to capture all machines.

In this scenario, we will still need one for MAPI and one for CAPI right?

I think having the ability to restrict to a particular namespace if required is a reasonable ask, but maybe isn't a blocker for this PR going forward

Reply:
thanks. agree that it isn't a blocker for this PR 👍